File: core/src/ru/myitschool/dcrawler/entities/skills/ShieldBash.java
package ru.myitschool.dcrawler.entities.skills;

import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.utils.Array;
import ru.myitschool.dcrawler.effects.ImmobilizedEffect;
import ru.myitschool.dcrawler.entities.Entity;
import ru.myitschool.dcrawler.math.DiceAction;
import ru.myitschool.dcrawler.math.MathAction;
import ru.myitschool.dcrawler.entities.skills.action.SimpleCombinedAction;
import ru.myitschool.dcrawler.entities.skills.patterns.EnemyTargetPattern;
import ru.myitschool.dcrawler.entities.skills.patterns.TargetPattern;
import ru.myitschool.dcrawler.entities.skills.targeting.EnemyDisplayer;
import ru.myitschool.dcrawler.entities.skills.targeting.TargetDisplayer;
import ru.myitschool.dcrawler.utils.Utils;

/**
 * Created by Voyager on 02.12.2017.
 */
public class ShieldBash extends Skill {

    MathAction attackAction = new DiceAction(6);
    int turns = 1;
    int distance = 1;

    public ShieldBash(Entity doer) {
        super(doer);
        addPlayContainer().getEnemyPlay().addAction(new SimpleCombinedAction(this, attackAction) {
            @Override
            public void successEffect(Target target, int damage, FloatingDamageMark mark) {
                Entity entity = target.getEntity();
                Utils.pushEntity(getDoer(), target.getEntity(), distance);
                entity.addEffect(new ImmobilizedEffect(entity, turns));
            }
        });
    }

    @Override
    public void maintainDisplayers(Array<TargetDisplayer> displayers) {
        displayers.add(new EnemyDisplayer());
    }

    @Override
    public TargetPattern getPattern() {
        return new EnemyTargetPattern(this);
    }

    @Override
    public Texture getIcon() {
        return new Texture("shield_bash.png");
    }

    @Override
    public String getName() {
        return "Shield Bash";
    }

    @Override
    public String getDescription() {
        return "You bashes target with your shield immobilizes it for " + turns + " turn(-s), trows it back for " + distance + " cell(-s) and deal " + attackAction.getDescription() + " damage";
    }

    @Override
    public int getSkillAccuracyBonus() {
        return 0;
    }

    @Override
    public int getRange() {
        return 1;
    }

    @Override
    public int getDistanceMin() {
        return 1;
    }

    @Override
    public int getDistanceMax() {
        return 1;
    }

    @Override
    public int getTargetCountMax() {
        return 1;
    }

    @Override
    public int getCooldownMax() {
        return 3;
    }

    @Override
    public int getType() {
        return SKILL_TYPE_COOLDOWN_DICE;
    }

    @Override
    public int getTargetType() {
        return SKILL_TARGET_TYPE_ENEMY;
    }

    @Override
    public boolean isCheckAllTargets() {
        return false;
    }

    @Override
    public boolean isMarkEverything() {
        return false;
    }

    @Override
    public boolean isMark() {
        return true;
    }

    @Override
    public boolean isObstruct() {
        return true;
    }

    @Override
    public boolean isWallTargets() {
        return false;
    }
}

File: core/src/com/wolg_vlad/dcrawler/entities/skills/ShieldBash.kt
package com.wolg_vlad.dcrawler.entities.skills

import com.badlogic.gdx.graphics.Texture
import com.badlogic.gdx.utils.Array
import com.wolg_vlad.dcrawler.effects.ImmobilizedEffect
import com.wolg_vlad.dcrawler.entities.Entity
import com.wolg_vlad.dcrawler.entities.skills.action.SimpleCombinedAction
import com.wolg_vlad.dcrawler.entities.skills.patterns.EnemyTargetPattern
import com.wolg_vlad.dcrawler.entities.skills.patterns.TargetPattern
import com.wolg_vlad.dcrawler.entities.skills.targeting.EnemyDisplayer
import com.wolg_vlad.dcrawler.entities.skills.targeting.TargetDisplayer
import com.wolg_vlad.dcrawler.math.DiceAction
import com.wolg_vlad.dcrawler.math.MathAction
import com.wolg_vlad.dcrawler.utils.Utils

/**
 * Created by Voyager on 02.12.2017.
 */
class ShieldBash(doer: Entity) : Skill(doer) {
    var attackAction: MathAction = DiceAction(6)
    var turns = 1
    var distance = 1

    override val displayers: kotlin.Array<TargetDisplayer> = arrayOf(EnemyDisplayer())
    override val pattern: TargetPattern = EnemyTargetPattern(this)
    override val icon: Texture = Texture("shield_bash.png")
    override val name: String = "Shield Bash"
    override val description: String = "You bashes target with your shield immobilizes it for $turns turn(-s), " +
            "throws it back for $distance cell(-s) and deal ${attackAction.description} damage"
    override val skillAccuracyBonus: Int = 0
    override val range: Int = 1
    override val distanceMin: Int = 1
    override val distanceMax: Int = 1
    override val targetCountMax: Int = 1
    override val cooldownMax: Int = 3
    override val type: Int = SKILL_TYPE_COOLDOWN_DICE
    override val targetType: Int = SKILL_TARGET_TYPE_ENEMY
    override val isCheckAllTargets: Boolean = false
    override val isMarkEverything: Boolean = false
    override val isMark: Boolean = true
    override val isObstruct: Boolean = true
    override val isWallTargets: Boolean = false

    init {
        addPlayContainer().enemyPlay.addAction(object : SimpleCombinedAction(this, attackAction) {
            public override fun successEffect(target: Target, damage: Int, mark: FloatingDamageMark) {
                val entity = target.entity
                Utils.pushEntity(doer, target.entity, distance)
                entity!!.addEffect(ImmobilizedEffect(entity, turns))
            }
        })
    }
}